File: tutorials/tutorial_10_metadata_stream/python/main.py
 
########################################################################
#
# Copyright (c) 2023, STEREOLABS.
#
# All rights reserved.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
########################################################################

import pyzed.sl as sl
import pyzed.sl_iot as sliot
import cv2

colors = [
    [232, 176, 244],
    [175, 208, 25],
    [102, 205, 105],
    [185, 0, 255],
    [99, 107, 252],
    [252, 225, 8],
    [167, 130, 141],
    [194, 72, 113]
 ]

def cvt(pt, scale):
    return[pt[0] * scale[0], pt[1] * scale[1]]


def main():
    global colors

    # Initialize the connection to ZED Hub
    status_iot = sliot.HubClient.connect("skeleton_tutorial")
    if status_iot != sliot.STATUS_CODE.SUCCESS:
        print("Initialization error ", status_iot)
        exit(1)

    init_params = sl.InitParameters()
    init_params.camera_resolution = sl.RESOLUTION.HD720
    init_params.depth_mode = sl.DEPTH_MODE.ULTRA
    init_params.coordinate_system = sl.COORDINATE_SYSTEM.RIGHT_HANDED_Y_UP
    init_params.coordinate_units = sl.UNIT.METER
    init_params.camera_fps = 20
    
    # Open the ZED camera
    zed = sl.Camera()
    err = zed.open(init_params)
    if err != sl.ERROR_CODE.SUCCESS:
        print("Error:", err)
        zed.close()
    
    # Register the camera once it's open
    update_params = sliot.UpdateParameters()
    status_iot = sliot.HubClient.register_camera(zed, update_params)
    if status_iot != sliot.STATUS_CODE.SUCCESS:
        print("Camera registration error:", status_iot)
        exit(1)
    
    # Enable Position tracking (mandatory for object detection)
    print("Enable Position Tracking Module")
    track_params = sl.PositionalTrackingParameters()
    track_params.enable_pose_smoothing = True
    track_params.set_as_static = False
    track_params.set_floor_as_origin = True
    err = zed.enable_positional_tracking(track_params)
    if err != sl.ERROR_CODE.SUCCESS:
        print("Error:", err)
        print("Exit program.")
        zed.close()
        exit(1)
    
    # Enable the Body Tracking module
    print("Enable Body Tracking Module")
    body_track_params = sl.BodyTrackingParameters()
    body_track_params.enable_tracking = True
    body_track_params.enable_body_fitting = False; # smooth skeletons moves
    body_track_params.body_format = sl.BODY_FORMAT.BODY_38
    body_track_params.detection_model = sl.BODY_TRACKING_MODEL.HUMAN_BODY_ACCURATE
    err = zed.enable_body_tracking(body_track_params)
    if err != sl.ERROR_CODE.SUCCESS:
        print("Error:", err)
        print("Exit program.")
        zed.close()
        exit(1)
        
    # Runtime parameters
    rt_params = sl.RuntimeParameters()
    rt_params.measure3D_reference_frame = sl.REFERENCE_FRAME.WORLD

    # Body Tracking runtime parameters
    body_track_rt_params = sl.BodyTrackingRuntimeParameters()
    body_track_rt_params.detection_confidence_threshold = 50
    
    # Image
    image = sl.Mat(1280, 720, sl.MAT_TYPE.U8_C4)
    cv_image = image.get_data()

    # 2D Drawing helpers
    camera_resolution = zed.get_camera_information().camera_configuration.resolution
    img_scale = [cv_image.shape[1] / camera_resolution.width, cv_image.shape[0] / camera_resolution.height]

    # Bodies to be streamed to ZED Hub
    bodies = sl.Bodies()

    # Main loop
    while True:
        # Grab a new frame from the ZED
        status_zed = zed.grab()

        if status_zed != sl.ERROR_CODE.SUCCESS:
            break

        # Retrieve left image
        zed.retrieve_image(image, sl.VIEW.LEFT)

        # Retrieve bodies
        zed.retrieve_bodies(bodies, body_track_rt_params)

        # Draw 2D skeletons
        for body in bodies.body_list:
            if body.tracking_state == sl.OBJECT_TRACKING_STATE.OK:
                color = colors[body.id % len(colors)]
                
                # Skeleton joints
                for kp in body.keypoint_2d:
                    cv_kp = cvt(kp, img_scale)
                
                    if cv_kp[0] < cv_image.shape[1] and cv_kp[1] < cv_image.shape[0]:
                        cv2.circle(cv_image, (int(cv_kp[0]), int(cv_kp[1])), 3, color, -1)

                # Skeleton bones
                for limb in sl.BODY_38_BONES:
                    kp_a = cvt(body.keypoint_2d[sl.get_idx_34(limb[0])], img_scale)
                    kp_b = cvt(body.keypoint_2d[sl.get_idx_34(limb[1])], img_scale)

                    if 0 < kp_a[0] < cv_image.shape[1] and 0 < kp_a[1] < cv_image.shape[0] and 0 < kp_b[0] < cv_image.shape[1] and 0 < kp_b[1] < cv_image.shape[0]:
                        cv2.line(cv_image, (int(kp_a[0]), int(kp_a[1])), (int(kp_b[0]), int(kp_b[1])), color, 1)

        # Always update Hub at the end of the grab loop to stream data to ZED Hub
        # Update the video stream/recording
        sliot.HubClient.update(zed, image)

        # Update the sl.Bodies stream
        sliot.HubClient.update_bodies(zed, bodies)
    
    # Handling camera error
    if status_zed != sl.ERROR_CODE.SUCCESS:
        sliot.HubClient.send_log("Grab failed, restarting camera. " + str(status_zed), sliot.LOG_LEVEL.ERROR)
        zed.close()
        sl.Camera.reboot(zed.get_camera_information().serial_number)
    
    # Close the camera
    elif zed.is_opened():
        zed.close()

    # Close the communication with ZED Hub properly.
    status_iot = sliot.HubClient.disconnect()
    if status_iot != sliot.STATUS_CODE.SUCCESS:
        print("Terminating error ", status_iot)
        exit(1)


if __name__ == "__main__":
    main()

File: tutorials/tutorial_10_metadata_stream/cpp/main.cpp
///////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2023, STEREOLABS.
//
// All rights reserved.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
///////////////////////////////////////////////////////////////////////////

#include <sl_iot/HubClient.hpp>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace sl;
using namespace sl_iot;
using json = sl_iot::json;

template<typename T>
inline cv::Point2f cvt(T pt, sl::float2 scale) {
    return cv::Point2f(pt.x * scale.x, pt.y * scale.y);
}

cv::Mat slMat2cvMat(Mat &input)
{
    // Mapping between MAT_TYPE and CV_TYPE
    int cv_type = -1;
    switch (input.getDataType())
    {
        case MAT_TYPE::F32_C1:
            cv_type = CV_32FC1;
            break;
        case MAT_TYPE::F32_C2:
            cv_type = CV_32FC2;
            break;
        case MAT_TYPE::F32_C3:
            cv_type = CV_32FC3;
            break;
        case MAT_TYPE::F32_C4:
            cv_type = CV_32FC4;
            break;
        case MAT_TYPE::U8_C1:
            cv_type = CV_8UC1;
            break;
        case MAT_TYPE::U8_C2:
            cv_type = CV_8UC2;
            break;
        case MAT_TYPE::U8_C3:
            cv_type = CV_8UC3;
            break;
        case MAT_TYPE::U8_C4:
            cv_type = CV_8UC4;
            break;
        default:
            break;
    }

    return cv::Mat(input.getHeight(), input.getWidth(), cv_type, input.getPtr<sl::uchar1>(MEM::CPU));
}

vector<cv::Scalar> colors = {
    cv::Scalar(232, 176, 244, 255),
    cv::Scalar(175, 208, 25, 255),
    cv::Scalar(102, 205, 105, 255),
    cv::Scalar(185, 0, 255, 255),
    cv::Scalar(99, 107, 252, 255),
    cv::Scalar(252, 225, 8, 255),
    cv::Scalar(167, 130, 141, 255),
    cv::Scalar(194, 72, 113, 255)
};

int main(int argc, char **argv)
{
    // Create camera object
    std::shared_ptr<sl::Camera> p_zed;
    p_zed.reset(new sl::Camera());

    // Initialize the connection to ZED Hub
    STATUS_CODE status_iot = HubClient::connect("skeleton_app");
    if (status_iot != STATUS_CODE::SUCCESS)
    {
        std::cout << "HubClient " << status_iot << std::endl;
        exit(EXIT_FAILURE);
    }

    // Open the ZED camera
    sl::InitParameters init_params;
    init_params.camera_resolution = RESOLUTION::HD720;
    init_params.depth_mode = DEPTH_MODE::ULTRA;
    init_params.coordinate_system = COORDINATE_SYSTEM::RIGHT_HANDED_Y_UP;
    init_params.coordinate_units = UNIT::METER;
    init_params.depth_stabilization = 20;

    sl::ERROR_CODE status_zed = p_zed->open(init_params);
    if (status_zed != ERROR_CODE::SUCCESS)
    {
        HubClient::sendLog("Camera initialization error : " + std::string(toString(status_zed)), LOG_LEVEL::ERROR);
        exit(EXIT_FAILURE);
    }

    // Register the camera once it's open
    UpdateParameters update_params;
    status_iot = HubClient::registerCamera(p_zed, update_params);
    if (status_iot != STATUS_CODE::SUCCESS) {
        std::cout << "Camera registration error " << status_iot << std::endl;
        exit(EXIT_FAILURE);
    }

    // Enable Position tracking (mandatory for object detection)
    std::cout << "Enable Position Tracking Module" << std::endl;
    sl::PositionalTrackingParameters track_params;
    track_params.enable_pose_smoothing = true;
    track_params.set_as_static = false;
    track_params.set_floor_as_origin = true;
    auto zed_error = p_zed->enablePositionalTracking(track_params);
    if (zed_error != ERROR_CODE::SUCCESS)
    {
        std::cout << sl::toVerbose(zed_error) << "\nExit program." << std::endl;
        p_zed->close();
        exit(EXIT_FAILURE);
    }

    // Enable the Body Tracking module
    std::cout << "Enable Body Tracking Module" << std::endl;
    sl::BodyTrackingParameters body_track_params;
    body_track_params.enable_tracking = true;
    body_track_params.enable_body_fitting = false; // smooth skeletons moves
    body_track_params.body_format = BODY_FORMAT::BODY_38;
    body_track_params.detection_model = BODY_TRACKING_MODEL::HUMAN_BODY_ACCURATE;
    zed_error = p_zed->enableBodyTracking(body_track_params);
    if (zed_error != ERROR_CODE::SUCCESS)
    {
        std::cout << sl::toVerbose(zed_error) << "\nExit program." << std::endl;
        p_zed->close();
        exit(EXIT_FAILURE);
    }

    // Runtime parameters
    RuntimeParameters rt_params;
    rt_params.measure3D_reference_frame = REFERENCE_FRAME::WORLD;

    // Body Tracking runtime parameters
    BodyTrackingRuntimeParameters body_track_rt_params;
    body_track_rt_params.detection_confidence_threshold = 50;

    // Images
    sl::Mat image(1280, 720, MAT_TYPE::U8_C4);
    cv::Mat cv_image;
    cv_image = slMat2cvMat(image);

    // 2D Drawing helpers
    Resolution camera_resolution = p_zed->getCameraInformation().camera_configuration.resolution;
    sl::float2 img_scale((float)cv_image.cols / (float)camera_resolution.width, (float)cv_image.rows / (float)camera_resolution.height);
    cv::Rect roi_render(0, 0, cv_image.cols, cv_image.rows);

    // Bodies to be streamed to ZED Hub
    sl::Bodies bodies;

    // Main loop
    while (true)
    {
        // Grab a new frame from the ZED
        status_zed = p_zed->grab(rt_params);
        if (status_zed != ERROR_CODE::SUCCESS)
            break;

        // Retrieve left image
        p_zed->retrieveImage(image, VIEW::LEFT, MEM::CPU, image.getResolution());
        
        // Retrieve bodies
        p_zed->retrieveBodies(bodies, body_track_rt_params);

        // Draw 2D skeletons
        for (int i = 0; i < bodies.body_list.size(); i++)
        {
            sl::BodyData& body = (bodies.body_list[i]);
            if (body.tracking_state == sl::OBJECT_TRACKING_STATE::OK)
            {
                cv::Scalar color = colors[body.id % colors.size()];

                // Skeleton joints
                for (auto& kp : body.keypoint_2d)
                {
                    auto cv_kp = cvt(kp, img_scale);
                    if (roi_render.contains(cv_kp))
                        cv::circle(cv_image, cv_kp, 3, color, -1);
                }

                // Skeleton bones
                for (auto& limb : BODY_38_BONES)
                {
                    auto kp_a = cvt(body.keypoint_2d[getIdx(limb.first)], img_scale);
                    auto kp_b = cvt(body.keypoint_2d[getIdx(limb.second)], img_scale);
                    if (roi_render.contains(kp_a) && roi_render.contains(kp_b))
                        cv::line(cv_image, kp_a, kp_b, color, 1);                                
                }
            }
        }

        // Always update Hub at the end of the grab loop to stream data to ZED Hub
        // Update the video stream/recording
        HubClient::update(p_zed, image);

        // Update the sl::Bodies stream
        HubClient::update(p_zed, bodies);
    }

    // Clear objects
    image.free();
    bodies.body_list.clear();
    
    // Handling camera error
    if (status_zed != ERROR_CODE::SUCCESS)
    {
        HubClient::sendLog("Grab failed, restarting camera. " + std::string(toString(status_zed)), LOG_LEVEL::ERROR);
        p_zed->close();
        sl::ERROR_CODE e = sl::Camera::reboot(p_zed->getCameraInformation().serial_number);
    }

    // Close the camera
    if (p_zed->isOpened())
        p_zed->close();

    status_iot = HubClient::disconnect();
    if (status_iot != STATUS_CODE::SUCCESS)
    {
        std::cout << "Terminating error " << status_iot << std::endl;
        exit(EXIT_FAILURE);
    }

    return 0;
}
